public/app/features/alerting/state/AlertingQueryRunner.ts

```diff
@@ -0,0 +1,208 @@
+import { merge, Observable, of, ReplaySubject, timer, Unsubscribable } from 'rxjs';
+import { catchError, finalize, map, mapTo, share, takeUntil } from 'rxjs/operators';
+import { v4 as uuidv4 } from 'uuid';
+import {
+  compareArrayValues,
+  compareDataFrameStructures,
+  dataFrameFromJSON,
+  DataFrameJSON,
+  getDefaultTimeRange,
+  LoadingState,
+  PanelData,
+  rangeUtil,
+  TimeRange,
+} from '@grafana/data';
+import { BackendSrvRequest, FetchResponse, toDataQueryError } from '@grafana/runtime';
+import { BackendSrv, getBackendSrv } from 'app/core/services/backend_srv';
+import { preProcessPanelData } from 'app/features/query/state/runRequest';
+import { GrafanaQuery } from 'app/types/unified-alerting-dto';
+import { getTimeRangeForExpression } from '../unified/utils/timeRange';
+import { isExpressionQuery } from 'app/features/expressions/guards';
+
+export interface AlertingQueryResult {
+  frames: DataFrameJSON[];
+}
+
+export interface AlertingQueryResponse {
+  results: Record<string, AlertingQueryResult>;
+}
+export class AlertingQueryRunner {
+  private subject: ReplaySubject<Record<string, PanelData>>;
+  private subscription?: Unsubscribable;
+  private lastResult: Record<string, PanelData>;
+
+  constructor(private backendSrv = getBackendSrv()) {
+    this.subject = new ReplaySubject(1);
+    this.lastResult = {};
+  }
+
+  get(): Observable<Record<string, PanelData>> {
+    return this.subject.asObservable();
+  }
+
+  run(queries: GrafanaQuery[]) {
+    if (queries.length === 0) {
+      const empty = initialState(queries, LoadingState.Done);
+      return this.subject.next(empty);
+    }
+
+    this.subscription = runRequest(this.backendSrv, queries).subscribe({
+      next: (dataPerQuery) => {
+        const nextResult: Record<string, PanelData> = {};
+
+        for (const [refId, data] of Object.entries(dataPerQuery)) {
+          const previous = this.lastResult[refId];
+          nextResult[refId] = setStructureRevision(data, previous);
+        }
+
+        this.lastResult = nextResult;
+        this.subject.next(this.lastResult);
+      },
+      error: (error) => console.error('PanelQueryRunner Error', error),
+    });
+  }
+
+  cancel() {
+    if (!this.subscription) {
+      return;
+    }
+    this.subscription.unsubscribe();
+
+    const nextResult: Record<string, PanelData> = {};
+    let loading = false;
+
+    for (const [refId, data] of Object.entries(this.lastResult)) {
+      if (data.state === LoadingState.Loading) {
+        loading = true;
+      }
+
+      nextResult[refId] = {
+        ...data,
+        state: LoadingState.Done,
+      };
+    }
+
+    if (loading) {
+      this.subject.next(nextResult);
+    }
```

Comment:
This does not cancel the runRequest observable, right? You could have a takeUntil in runRequest that would cancel it too.

Reply:
aha, now I get your point.